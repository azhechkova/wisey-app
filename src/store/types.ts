import { TrackVideoType } from '~/types';

export type RootState = {
  app: {
    activeLesson: string;
    videoProgress: TrackVideoType[];
  };
};
