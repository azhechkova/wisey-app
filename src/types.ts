export type LessonType = {
  id: string;
  title: string;
  duration: number;
  order: number;
  type: 'video';
  status: 'unlocked' | 'locked';
  link: string;
  previewImageLink: string;
  meta: object | null;
};

export type CourseType = {
  id: string;
  title: string;
  tags: string[];
  launchDate?: string;
  status: 'launched';
  description: string;
  duration: number;
  lessonsCount?: number;
  lessons: LessonType[];
  containsLockedLessons: boolean;
  previewImageLink: string;
  rating: number;
  meta?: {
    slug?: string;
    skills?: string[];
    courseVideoPreview?: {
      link?: string;
      previewImageLink?: string;
      duration?: number;
    };
  } | null;
};
