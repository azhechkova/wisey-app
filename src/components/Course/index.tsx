import React from 'react';
import {
  CardContent,
  Typography,
  Card,
  CardActions,
  Button,
  List,
  ListItem,
} from '@mui/material';
import { useNavigate } from 'react-router-dom';

import useStyles from './styles';
import { CourseType } from '../../types';

import Tag from '../Tag';
import CourseVideo from '../CourseVideo';

interface CourseProps {
  course: CourseType;
}

const Course = ({ course }: CourseProps): JSX.Element => {
  const { classes } = useStyles();
  const navigate = useNavigate();

  const onNavigate = (): void => navigate(`/courses/${course.id}`);

  const imageLink = `${course.previewImageLink}/cover.webp`;

  return (
    <Card className={classes.card}>
      <CourseVideo
        src={course?.meta?.courseVideoPreview?.link}
        width="352"
        height="198"
        controls={false}
        poster={imageLink}
        muted
        playOnHover
      />
      <CardContent className={classes.content}>
        <Typography gutterBottom variant="h5" component="h2">
          {course.title}
        </Typography>
        <Typography variant="body2" color="text.primary">
          {course.description}
        </Typography>
        <List>
          <ListItem className={classes.listItem}>
            Rating: {course.rating}
          </ListItem>
          <ListItem className={classes.listItem}>
            Lessons: {course.lessonsCount}
          </ListItem>
        </List>
        <CardActions>
          <Button variant="contained" size="small" onClick={onNavigate}>
            Learn More
          </Button>
        </CardActions>
        {course.tags.map(tag => (
          <Tag label={tag} key={tag} />
        ))}
      </CardContent>
    </Card>
  );
};

export default Course;
