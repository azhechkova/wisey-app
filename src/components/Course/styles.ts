import { makeStyles } from 'tss-react/mui';

const useStyles = makeStyles()(() => ({
  card: {
    maxWidth: 345,
    boxShadow:
      'rgba(24, 39, 75, 0.12) 0px 8px 18px -6px, rgba(24, 39, 75, 0.12) 0px 12px 42px -4px',
    borderRadius: '1em',
  },
  content: {
    padding: '1em',
  },
  listItem: {
    padding: '0.5em 0',
  },
}));

export default useStyles;
