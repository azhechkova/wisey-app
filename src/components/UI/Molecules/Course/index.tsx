import React, { useRef } from 'react';
import {
  CardContent,
  Box,
  Typography,
  Card,
  List,
  ListItem,
} from '@mui/material';
import { Link } from 'react-router-dom';
import SchoolIcon from '@mui/icons-material/School';
import Rating from '@mui/material/Rating';
import { CourseType } from '~/types';

import Tag from '~/components/UI/Atoms/Tag';
import CourseVideo from '~/components/UI/Atoms/CourseVideo';
import useStyles from './styles';

interface CourseProps {
  course: CourseType;
}

const Course = ({ course }: CourseProps): JSX.Element => {
  const { classes } = useStyles();
  const ref = useRef<HTMLVideoElement | null>(null);
  const imageLink = `${course.previewImageLink}/cover.webp`;

  return (
    <Link to={`/courses/${course.id}`} className={classes.root}>
      <Card className={classes.card}>
        <CourseVideo
          className={classes.video}
          src={course?.meta?.courseVideoPreview?.link}
          width="100%"
          height="198"
          controls={false}
          poster={imageLink}
          muted
          playOnHover
          videoRef={ref}
        />
        <CardContent className={classes.content}>
          <Typography fontWeight={700} variant="h6" component="h2">
            {course.title}
          </Typography>
          <Typography
            className={classes.description}
            variant="body2"
            color="text.primary"
          >
            {course.description}
          </Typography>
          <Box className={classes.info}>
            <Rating value={course.rating} />
            <Typography className={classes.lessonsCount}>
              <SchoolIcon /> {course.lessonsCount}
            </Typography>
          </Box>
          <List className={classes.list}>
            {course.meta?.skills?.length &&
              course.meta.skills.map(tag => (
                <ListItem key={tag} className={classes.listItem}>
                  <Tag label={tag} />
                </ListItem>
              ))}
          </List>
        </CardContent>
      </Card>
    </Link>
  );
};

export default Course;
