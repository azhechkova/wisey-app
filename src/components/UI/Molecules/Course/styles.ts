import { makeStyles } from 'tss-react/mui';

const useStyles = makeStyles()(theme => ({
  root: {
    textDecoration: 'none',
    color: theme.palette.text.primary,
    height: 'auto',
    width: '100%',
    '&:hover': {
      transform: 'scale(1.025)',
      transition: 'transform ease .5s',
    },
  },
  card: {
    height: '100%',
    boxShadow:
      'rgba(24, 39, 75, 0.12) 0px 8px 18px -6px, rgba(24, 39, 75, 0.12) 0px 12px 42px -4px',
    borderRadius: '1em',
    display: 'flex',
    flexDirection: 'column',
  },
  content: {
    padding: '1em',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    flexGrow: 1,
  },
  list: {
    padding: '0',
  },
  listItem: {
    padding: '0',
  },
  description: {
    lineHeight: '1.25rem',
    fontSize: '0.875rem',
    marginTop: '0.75rem',
  },
  info: {
    margin: '15px 0',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  video: {
    objectFit: 'cover',
  },
  lessonsCount: {
    display: 'flex',
    alignItems: 'center',
    gap: '5px',
  },
}));

export default useStyles;
