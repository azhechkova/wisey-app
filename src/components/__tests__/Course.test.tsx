import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { CourseType } from '../../types';

import Course from '../UI/Molecules/Course';

const course: CourseType = {
  id: '1',
  title: 'Mock course',
  tags: ['tag'],
  status: 'launched',
  description: 'Course description',
  duration: 100,
  lessons: [],
  containsLockedLessons: false,
  previewImageLink: '',
  rating: 1,
};

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: (): jest.Mock<void> => mockedUsedNavigate,
}));

describe('Course component', () => {
  test('renders properly', () => {
    const { container } = render(<Course course={course} />);

    expect(container).toMatchSnapshot();
  });
  test('has description and title', () => {
    render(<Course course={course} />);

    expect(screen.getByText(course.title)).toBeInTheDocument();
    expect(screen.getByText(course.description)).toBeInTheDocument();
  });

  test('navigates to the course page', () => {
    render(<Course course={course} />);

    fireEvent.click(screen.getByText('Learn More'));
    expect(mockedUsedNavigate).toHaveBeenCalledWith(`/courses/${course.id}`);
  });
});
