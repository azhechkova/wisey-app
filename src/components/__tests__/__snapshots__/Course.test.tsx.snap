// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course component renders properly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-111azgf-MuiPaper-root-MuiCard-root-card"
  >
    <video
      height="198"
      poster="/cover.webp"
      width="352"
    >
      <source
        type="application/x-mpegURL"
      />
    </video>
    <div
      class="MuiCardContent-root css-mman5-MuiCardContent-root-content"
    >
      <h2
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
      >
        Mock course
      </h2>
      <p
        class="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
      >
        Course description
      </p>
      <ul
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-ba5xb9-MuiListItem-root-listItem"
        >
          Rating: 
          1
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-ba5xb9-MuiListItem-root-listItem"
        >
          Lessons: 
        </li>
      </ul>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-11qr2p8-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Learn More
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-15d77bz-MuiChip-root-tag"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          tag
        </span>
      </div>
    </div>
  </div>
</div>
`;
